ceresdb/client_impl.go

```diff
@@ -4,86 +4,79 @@ package ceresdb
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"github.com/CeresDB/ceresdb-client-go/types"
 	"github.com/CeresDB/ceresdb-client-go/utils"
-	"github.com/CeresDB/ceresdbproto/go/ceresdbproto"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials/insecure"
-)
-
-const (
-	codeOk = 200
 )
 
 type clientImpl struct {
-	inner ceresdbproto.StorageServiceClient
-	conn  *grpc.ClientConn
+	rpcClient   *rpcClient
+	routeClient *routeClient
 }
 
-func newClient(endpoint string, opts *options) (CeresDBClient, error) {
-	conn, err := grpc.Dial(endpoint,
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
-		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(opts.RpcMaxRecvMsgSize)))
+func newClient(endpoint string, opts options) (CeresDBClient, error) {
+	rpcClient := newRpcClient(opts)
+	routeClient, err := newRouteClient(endpoint, rpcClient, opts)
 	if err != nil {
 		return nil, err
 	}
-
-	c := ceresdbproto.NewStorageServiceClient(conn)
-
 	return &clientImpl{
-		inner: c,
-		conn:  conn,
+		rpcClient:   rpcClient,
+		routeClient: routeClient,
 	}, nil
 }
 
 func (c *clientImpl) Query(ctx context.Context, req types.QueryRequest) (types.QueryResponse, error) {
-	queryRequest := &ceresdbproto.QueryRequest{
-		Metrics: req.Metrics,
-		Ql:      req.Ql,
+	if len(req.Metrics) == 0 {
+		return types.QueryResponse{}, errors.New("Null request metrics")
 	}
 
-	queryResponse, err := c.inner.Query(ctx, queryRequest)
+	routes, err := c.routeClient.RouteFor(req.Metrics)
 	if err != nil {
-		return types.QueryResponse{}, err
+		return types.QueryResponse{}, fmt.Errorf("Route metrics %v failed, err %v", req.Metrics, err)
 	}
-	if queryResponse.Header.Code != codeOk {
-		return types.QueryResponse{}, fmt.Errorf("query failed, code: %d, err: %s",
-			queryResponse.Header.Code, queryResponse.Header.Error)
+	for _, route := range routes {
+		queryResponse, err := c.rpcClient.Query(route.Endpoint, ctx, req)
+		if ceresdbErr, ok := err.(*types.CeresdbError); ok && ceresdbErr.ShouldRefreshRoute() {
```

Comment:
Why add `ShouldRefreshRoute()`?
`ShouldRefreshRoute` means refresh router not `ClearRouteFor`.